src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,218 @@
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
-            }
-            else if (errno == ENOENT)
-            {                
-                sMissingDevURandom = true;
-            }
+    return fd;
+}
+
+static void FillWithMrand(uint8_t *buffer, int32_t bufferLength)
+{
+    static bool sInitializedMRand;
+    size_t length = (size_t)bufferLength;
+
+    if (!sInitializedMRand)
+    {
+        // FIXME: There's not enough entropy in time(NULL) to seed the PRNG.
+        // Leverage other sources of entropy?
+        srand48(time(NULL));
+        sInitializedMRand = true;
+    }
+
+    while (length)
+    {
+        long num = lrand48();
+
+        if (length >= sizeof(num))
+        {
+            memcpy(buffer, &num, sizeof(num));
+
+            buffer += sizeof(num);
+            length -= sizeof(num);
+        }
+        else
+        {
+            memcpy(buffer, &num, length);
         }
+    }
+}
+
+#if HAVE_SYS_GETRANDOM
+static bool IsGetRandomAvailable(void)
+{
+    uint8_t buffer;
+    long r = syscall(SYS_getrandom, &buffer, 1, GRND_NONBLOCK);
+
+    if (r == 1)
+    {
+        return true;
+    }
+    if ((r == -1) && (errno != ENOSYS))
+    {
+        // At this point, we're only interested if getrandom(2) is available
+        // in this kernel.  If errno isn't ENOSYS, then we're (probably) good.
+        return true;
+    }
 
-        if (rand_des != -1)
+    return false;
+}
+
+static bool TryFillWithGetRandom(uint8_t* buffer, int32_t bufferLength)
+{
+    size_t length = (size_t)bufferLength;
+
+    do
+    {
+        long n = syscall(SYS_getrandom, buffer, length, GRND_NONBLOCK);
+
+        if (n == -1)
         {
-            int32_t offset = 0;
-            do
+            if (errno == EINTR)
             {
-                ssize_t n = read(rand_des, buffer + offset , (size_t)(bufferLength - offset));
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-
-                    assert(false && "read from /dev/urandom has failed");
-                    break;
-                }
-
-                offset += n;
+                continue;
             }
-            while (offset != bufferLength);
+
+            assert(false && "getrandom() has failed");
+            return false;
         }
+
+        buffer += n;
+        length -= (size_t)n;
     }
+    while(length);
 
-    if (!sInitializedMRand)
+    return true;
+}
+
+static void FillWithGetRandom(uint8_t* buffer, int32_t bufferLength)
+{
+    if (!TryFillWithGetRandom(buffer, bufferLength))
     {
-        srand48(time(NULL));
-        sInitializedMRand = true;
+        FillWithMrand(buffer, bufferLength);
     }
+}
+#endif // HAVE_SYS_GETRANDOM
 
-    // always xor srand48 over the whole buffer to get some randomness
```

Comment:
Why did you get rid of this behavior?

Reply:
Because it's better to trust urandom, that was properly seeded (and is the product of various other sources of random data already), rather than using another PRNG that has been seeded with `time(NULL)` to flip some of those bits.

The previous behavior not only did unnecessary work, but was also questionable.